### PR TITLE
GH-437 Improve schedulers

### LINKED ISSUE #437: Clean up shchedulers
LiteCommands provides a default scheduler [SchedulerExecutorPoolImpl](https://github.com/Rollczi/LiteCommands/blob/0f4af80ffb59d0c5b55b9f431f2971c63a30d187/litecommands-core/src/dev/rollczi/litecommands/scheduler/SchedulerExecutorPoolImpl.java) but it may not be the best way to control threads on platforms such as Fabric BungeeCord Fabric Sponge Velocity

### PR DESCRIPTION
**Pull request by BlackBaroness**
Sorry for being late! Will finish soon! Closes #437.

I am currently reviewing and improving schedulers for all platforms:
- [x] Core
- [x] Bukkit
- [x] BungeeCord
- [x] Sponge
- [x] Velocity
- [x] Fabric
- [x] Jda
- [x] Minestom

### About BungeeCord

> BungeeCord's own scheduler is not efficient because it creates an unlimited cached thread pool for
> each plugin and just uses it:
> https://github.com/SpigotMC/BungeeCord/blob/master/proxy/src/main/java/net/md_5/bungee/scheduler/BungeeScheduler.java
> https://github.com/SpigotMC/BungeeCord/blob/master/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java
> 
> It also just blocks the threads if a task with a delay is received (by using Thread.sleep),
> so the thread won't be doing anything useful even if there is more tasks available

I decided to use the Core scheduler instead, since it's more efficient by at least providing a max cap

### About Velocity

> Velocity's own scheduler is very similar to BungeeCord's one
> Each plugin has it's own uncapped cached thread pool
> That doesn't look good
> https://github.com/PaperMC/Velocity/blob/dev/3.0.0/proxy/src/main/java/com/velocitypowered/proxy/plugin/loader/VelocityPluginContainer.java
> https://github.com/PaperMC/Velocity/blob/dev/3.0.0/api/src/main/java/com/velocitypowered/api/plugin/PluginContainer.java

I decided to use the Core scheduler instead, since it's more efficient by at least providing a max cap

### About Bukkit

Looks good, only small improvements were made

### About Minestom

Don't see anything to improve, it's fine. Also the Minestom's own scheduler is good, so no need to use our own

### About Sponge

Sponge provides a scheduler similar to Bukkit, so I implemented a wrapper for one

### About JDA

I don't think there is a specific implementation for JDA, so we should just use the one from Core

### About Fabric

Was implemented by @huanmeng-qwq

### CODE REVIEW

**Comment by huanmeng-qwq** [comment]
@BlackBaroness Hi, I noticed that you can't solve the fabric platform. Since I don't have permission to submit changes directly to the pr, I will create a new [PR](https://github.com/BlackBaroness/LiteCommands/pull/1) and submit it to your fork warehouse.

**Comment by BlackBaroness** [comment]
> @BlackBaroness Hi, I noticed that you can't solve the fabric platform. Since I don't have permission to submit changes directly to the pr, I will create a new [PR](https://github.com/BlackBaroness/LiteCommands/pull/1) and submit it to your fork repository.

Thanks, I merged it

File: litecommands-core/src/dev/rollczi/litecommands/scheduler/SchedulerExecutorPoolImpl.java [hunk 1]
```diff
@@ -13,79 +18,137 @@
...
 
             return thread;
         };
 
-        this.asyncExecutor = pool < 0 ? Executors.newCachedThreadPool(factory) : Executors.newFixedThreadPool(pool, factory);
+        this.asyncExecutor = Executors.newScheduledThreadPool(
+            Math.max(pool, Runtime.getRuntime().availableProcessors()) / 2,
+            factory
+        );
+        this.closeAsyncExecutorOnShutdown = true;
```
Comment (Rollczi) [inline litecommands-core/src/dev/rollczi/litecommands/scheduler/SchedulerExecutorPoolImpl.java:53]: Do not change the way this class works and do not adapt its operation for unit tests. If you want you can create a new implementation but do not violate the current API and its implementation.

Your implementation splits 2 pools, both of which have multiple threads, it has nothing to do with the "main" thread and can lead to concurrency in unexpected places.

**Review by Rollczi — CHANGES_REQUESTED** [review]

File: litecommands-core/src/dev/rollczi/litecommands/scheduler/SchedulerExecutorPoolImpl.java (hunk 1 above)
Comment (Rollczi) [reply litecommands-core/src/dev/rollczi/litecommands/scheduler/SchedulerExecutorPoolImpl.java:53]: Let's do it this way, let's keep `SchedulerExecutorPoolImpl` in the form it was. Let's add a separate implementation: `SchedulerFullAsyncImpl` :P

File: litecommands-core/src/dev/rollczi/litecommands/shared/ThrowingRunnable.java [hunk 2]
```diff
@@ -4,5 +4,4 @@
 public interface ThrowingRunnable<E extends Throwable> {
 
     void run() throws E;
-
```
Comment (Rollczi) [inline litecommands-core/src/dev/rollczi/litecommands/shared/ThrowingRunnable.java:7]: unrelated codestyle changes

File: litecommands-bukkit/src/dev/rollczi/litecommands/bukkit/BukkitSchedulerImpl.java [hunk 3]
```diff
@@ -48,8 +49,7 @@ private <T> CompletableFuture<T> supplySync(SchedulerPoll type, ThrowingSupplier
 
         if (delay.isZero()) {
             bukkitScheduler.runTask(plugin, () -> tryRun(type, future, supplier));
-        }
-        else {
+        } else {
```
Comment (Rollczi) [inline litecommands-bukkit/src/dev/rollczi/litecommands/bukkit/BukkitSchedulerImpl.java:52]: unrelated codestyle changes

File: litecommands-bukkit/src/dev/rollczi/litecommands/bukkit/BukkitSchedulerImpl.java [hunk 4]
```diff
@@ -72,12 +71,10 @@ private <T> CompletableFuture<T> supplyAsync(SchedulerPoll type, ThrowingSupplie
     private <T> CompletableFuture<T> tryRun(SchedulerPoll type, CompletableFuture<T> future, ThrowingSupplier<T, Throwable> supplier) {
         try {
             future.complete(supplier.get());
-        }
-        catch (Throwable throwable) {
+        } catch (Throwable throwable) {
```
Comment (Rollczi) [inline litecommands-bukkit/src/dev/rollczi/litecommands/bukkit/BukkitSchedulerImpl.java:74]: unrelated codestyle changes

File: litecommands-annotations/test/dev/rollczi/litecommands/annotations/async/AsyncArgumentsInitializeTest.java [hunk 5]
```diff
@@ -50,7 +50,6 @@ void test() {
         List<CompletableFuture<AssertExecute>> futures = new ArrayList<>();
         for (int i = 0; i < 50; i++) {
             TestPlatform platform = LiteCommandsTestFactory.startPlatform(config -> config
-                .scheduler(new SchedulerExecutorPoolImpl("lite-commands", 10))
```
Comment (Rollczi) [inline litecommands-annotations/test/dev/rollczi/litecommands/annotations/async/AsyncArgumentsInitializeTest.java:53]: Please revert it, this unit test is used to check a case with the specific scheduler.

File: litecommands-annotations/test/dev/rollczi/litecommands/annotations/async/AsyncCommandTest.java [hunk 6]
```diff
@@ -24,11 +23,19 @@
...
 
+    private static final String mainThreadName = "main_thread";
+    private static final String asyncThreadName = "async_thread";
+
     private static final int DELAY = 500;
     private static final int MARGIN = 200;
 
     static LiteTestConfig config = builder -> builder
-        .scheduler(new SchedulerExecutorPoolImpl("test", 1))
+        .scheduler(
```
Comment (Rollczi) [inline litecommands-annotations/test/dev/rollczi/litecommands/annotations/async/AsyncCommandTest.java:33]: Please revert it, this unit test is used to check a case with the specific scheduler.

File: litecommands-annotations/test/dev/rollczi/litecommands/annotations/async/ParallelAsyncCommandTest.java [hunk 7]
```diff
@@ -24,7 +24,6 @@ public class ParallelAsyncCommandTest extends LiteTestSpec {
     private static final int DELAY = 100;
 
     static LiteTestConfig config = builder -> builder
-        .scheduler(new SchedulerExecutorPoolImpl("test"))
```
Comment (Rollczi) [inline litecommands-annotations/test/dev/rollczi/litecommands/annotations/async/ParallelAsyncCommandTest.java:27]: Please revert it, this unit test is used to check a case with the specific scheduler.

File: litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    private class LaterTask<T> implements Runnable {
+        private final ThrowingSupplier<T, @NotNull Throwable> supplier;
+        private final CompletableFuture<T> future;
+
+        private final long end;
+
+        private LaterTask(Duration delay, ThrowingSupplier<T, @NotNull Throwable> supplier, CompletableFuture<T> future) {
+            this.supplier = supplier;
+            this.future = future;
+            end = System.currentTimeMillis() + delay.toMillis();
```
Comment (Rollczi) [inline litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java:95]: If you want to run the task later, please use scheduled executor.

**Review by Rollczi — CHANGES_REQUESTED** [review]

**Comment by Rollczi** [comment]
Using platform schedulers  (BungeeCord, Velocity) can improve performance by sharing unused threads in the cached pools

File: litecommands-core/src/dev/rollczi/litecommands/scheduler/SchedulerExecutorPoolImpl.java (hunk 1 above)
Comment (BlackBaroness) [reply litecommands-core/src/dev/rollczi/litecommands/scheduler/SchedulerExecutorPoolImpl.java:53]: ok sure

**Comment by BlackBaroness** [comment]
> Using platform schedulers (BungeeCord, Velocity) can improve performance by sharing unused threads in the cached pools

I believe uncapped cached pools dedicated to a single plugin only is a bad idea. Especially if you look on how delayed tasks were implemented - they create a thread and then call `Thread.sleep` inside it, so the pool can grow so much without even doing anything. If this pool was used by the entire proxy, of course it would be nice (like craftbukkit does), but in this case... I feel like the BungeeCord Scheduler is a mistake by design

Velocity is slightly better, but still has an uncapped pool. What if several players will spam async commands?

File: litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java (hunk 8 above)
Comment (BlackBaroness) [reply litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java:95]: @huanmeng-qwq could you please improve this?

Comment (huanmeng-qwq) [reply litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java:95]: > @huanmeng-qwq could you please improve this?

https://github.com/BlackBaroness/LiteCommands/pull/2

Comment (BlackBaroness) [reply litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java:95]: > > @huanmeng-qwq could you please improve this?
> 
> [BlackBaroness#2](https://github.com/BlackBaroness/LiteCommands/pull/2)

oh, sorry, I didn't notice this

**Comment by Rollczi** [comment]
> > Using platform schedulers (BungeeCord, Velocity) can improve performance by sharing unused threads in the cached pools
> 
> I believe an uncapped cached pools dedicated to a single plugin only is a bad idea. Especially if you look on how delayed tasks were implemented - they create a thread and then call `Thread.sleep` inside it, so the pool can grow so much without even doing anything. If this pool was used by the entire proxy, of course it would be nice (like craftbukkit does), but in this case... I feel like the BungeeCord Scheduler is a mistake by design
> 
> Velocity is slightly better, but still has an uncapped pool. What if several players will spam async commands?
> 
> Proxies are designed to handle thousands of players online concurrently, an uncapped pool can become a disaster when commands are expensive enough to make threads busy
> 
> Or, for BungeeCord... Delayed tasks are usually used by cooldown system. What if we have a `/login` command (what is something all players use at least once) with 2s cooldown? If 200-300 players joined after restart, everyone executed /login (what is async itself) - 200-300 threads were created and then 200-300 threads more created to make delayed tasks (as BungeeCord doesn't use scheduled pools). That's insane

Okay, this is actually true, let's use pool with limited threads.

File: litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+                CompletableFuture<T> future = new CompletableFuture<>();
+                //noinspection ResultOfMethodCallIgnored
+                getMainThreadExecutor().submit(() -> {
+                    try {
+                        future.complete(supplier.get());
+                    } catch (Throwable e) {
+                        future.completeExceptionally(e);
+                    }
+                });
+                return future.join();
```
Comment (Rollczi) [inline litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java:62]: this implementation doesn't look too good, the join blocks the thread

File: litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        public void run() throws Throwable {
+            if (resolve.equals(SchedulerPoll.MAIN)) {
+                try {
+                    getMainThreadExecutor().submit(() -> {
+                        try {
+                            runnable.run();
+                        } catch (Throwable e) {
+                            throw new RuntimeException(e);
+                        }
+                    }).join();
```
Comment (Rollczi) [inline litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java:88]: also in this place

**Comment by Rollczi** [comment]
I will try fix some issues

**Comment by Rollczi** [comment]
@BlackBaroness Thank you for your substantive comments, I implemented them into the old scheduler implementation. I also simplified the code for Farbric scheduler, what do you think, that should work? @huanmeng-qwq

**Comment by BlackBaroness** [comment]
@Rollczi, looks nice!

I noticed you created the `SchedulerMainThreadBased` class and I believe its naming is pretty strange, how about the `MainThreadBasedScheduler`?

Also, this new abstraction could be used for Sponge scheduler, I think.

**Comment by BlackBaroness** [comment]
Also, on Sponge we have both `client.scheduler` and `server.scheduler`. I am not sure which one to use, so used the server one. LiteCommands can be installed on clients without a server I guess, it doesn't really depend on server. But how do we specify which scheduler should we use?

We could make it like `if (server.available) server else client` on platform init, but what if both server and client are available? I am not sure if it's possible, maybe on singleplayer the server runs in completely different process

**Comment by Rollczi** [comment]
> I noticed you created the `SchedulerMainThreadBased` class and I believe its naming is pretty strange, how about the `MainThreadBasedScheduler`?

okay, I will rename it

> Also, this new abstraction could be used for Sponge scheduler, I think.

I don't think so, this will more complicated... let's follow the principe KISS (Keep it simple stupid)

**Comment by Rollczi** [comment]
> Also, on Sponge we have both `client.scheduler` and `server.scheduler`. I am not sure which one to use, so used the server one. LiteCommands can be installed on clients without a server I guess, it doesn't really depend on server. But how do we specify which scheduler should we use?
> 
> We could make it like `if (server.available) server else client` on platform init, but what if both server and client are available? I am not sure if it's possible, maybe on singleplayer the server runs in completely different process

user should use `LiteFabricFactory.server()` or `LiteFabricFactory.client()` depends on his implementation

File: litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            this.getMainThreadExecutor().submit(task);
+        } else {
+            this.scheduledExecutor.schedule(() -> this.getMainThreadExecutor().submit(task), delay.toMillis(), TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @Override
+    protected void runAsynchronous(Runnable task, Duration delay) {
+        if (delay.isZero()) {
+            this.getMainThreadExecutor().submit(task);
```
Comment (huanmeng-qwq) [inline litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java:36]:
```suggestion
            this.asyncExecutor.submit(task);
```

File: litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        } else {
+            this.scheduledExecutor.schedule(() -> this.getMainThreadExecutor().submit(task), delay.toMillis(), TimeUnit.MILLISECONDS);
```
Comment (huanmeng-qwq) [inline litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java:38]: This will cause tasks to be executed on the main game thread, and the meaning of asynchronousness will be lost.
```suggestion
            this.asyncExecutor.schedule(task, delay.toMillis(), TimeUnit.MILLISECONDS);
```

Comment (Rollczi) [reply litecommands-fabric/src/main/java/dev/rollczi/litecommands/fabric/FabricScheduler.java]: Thanks  for your review @huanmeng-qwq  I missed that

File: litecommands-sponge/src/dev/rollczi/litecommands/sponge/LiteSpongeFactory.java [hunk 13]
```diff
@@ -27,6 +27,12 @@ public static <B extends LiteCommandsBuilder<CommandCause, LiteSpongeSettings, B
         return (B) LiteCommandsFactory.builder(CommandCause.class, new SpongePlatform(plugin, settings)).selfProcessor((builder, internal) -> {
             MessageRegistry<CommandCause> messageRegistry = internal.getMessageRegistry();
 
+            if (game.isServerAvailable()) {
```
Comment (Rollczi) [inline litecommands-sponge/src/dev/rollczi/litecommands/sponge/LiteSpongeFactory.java:30]: Please, move this to constructor, (and extract to private method and throw when server and client is not available, ISE)

File: litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        asyncScheduler.submit(createTask(task, delay));
+    }
+
+    @Override
+    protected void log(Throwable throwable) {
+        pluginContainer.logger().error("Error completing command future:", throwable);
+    }
+
+    private Task createTask(Runnable task, Duration delay) {
+        Task.Builder builder = Task.builder().plugin(pluginContainer).execute(task);
```
Comment (Rollczi) [inline litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java:41]: Chain code style

File: litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+import dev.rollczi.litecommands.scheduler.AbstractMainThreadBasedScheduler;
+import org.spongepowered.api.scheduler.Task;
+import org.spongepowered.plugin.PluginContainer;
+
+import java.time.Duration;
+
+public class SpongeScheduler extends AbstractMainThreadBasedScheduler {
+
+    private final PluginContainer pluginContainer;
+    private final org.spongepowered.api.scheduler.Scheduler mainScheduler;
```
Comment (Rollczi) [inline litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java:12]: Can we use this class without local import?

**Review by Rollczi — COMMENTED** [review]
Thanks for next improvements @BlackBaroness

File: litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    protected void log(Throwable throwable) {
+        pluginContainer.logger().error("Error completing command future:", throwable);
+    }
+
+    private Task createTask(Runnable task, Duration delay) {
+        return Task.builder()
+            .plugin(pluginContainer)
+            .execute(task)
+            .delay(delay.isPositive() ? delay : Duration.ZERO)
```
Comment (Rollczi) [inline litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java:46]: we can put ZERO to delay method? This will work same such as without calling `delay()`?

File: litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    private Scheduler findMainScheduler(Game game) {
+        if (game.isServerAvailable()) {
+            return game.server().scheduler();
+        } else if (game.isClientAvailable()) {
+            return game.client().scheduler();
+        } else {
+            throw new IllegalStateException("Neither the server nor the client is available");
+        }
```
Comment (Rollczi) [inline litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java:57]:
```suggestion
        if (game.isServerAvailable()) {
            return game.server().scheduler();
        }
        
        if (game.isClientAvailable()) {
            return game.client().scheduler();
        }
        
        throw new IllegalStateException("Neither the server nor the client is available");
```

File: litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java (hunk 16 above)
Comment (BlackBaroness) [reply litecommands-sponge/src/dev/rollczi/litecommands/sponge/SpongeScheduler.java:46]: > we can put ZERO to delay method? This will work same such as without calling `delay()`?

yep, in fact, it is zero by default https://github.com/SpongePowered/Sponge/blob/7d3050f98615fde7c1822d9926883b5f471de127/src/main/java/org/spongepowered/common/scheduler/SpongeTask.java#L88

**Comment by Rollczi** [comment]
Okay thanks @BlackBaroness and @huanmeng-qwq for your help! ❤️
